optuna/integration/pytorch_distributed.py

```diff
@@ -0,0 +1,249 @@
+            return bool(self.delegate.should_prune())
+
+        # torch.bool seems to be the correct type, but the communication fails
+        # due to the RuntimeError.
+        return _call_and_communicate(func, torch.uint8)
+
+    def set_user_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self.delegate is not None
+            self.delegate.set_user_attr(key, value)
+        dist.barrier()
+
+    def set_system_attr(self, key: str, value: Any) -> None:
+
+        if dist.get_rank() == 0:
+            assert self.delegate is not None
+            self.delegate.set_system_attr(key, value)
+        dist.barrier()
+
+    @property
+    def number(self) -> int:
+        def func() -> int:
+
+            assert self.delegate is not None
+            return self.delegate.number
+
+        return _call_and_communicate(func, torch.int)
+
+    @property
+    def trial_id(self) -> int:
+
+        warnings.warn(
+            "The use of `TorchDistributedTrial.trial_id` is deprecated. "
+            "Please use `TorchDistributedTrial.number` instead.",
+            FutureWarning,
+        )
+        return self._trial_id
+
+    @property
+    def _trial_id(self) -> int:
+        def func() -> int:
+
+            assert self.delegate is not None
+            return self.delegate._trial_id
+
+        return _call_and_communicate(func, torch.int)
+
+    @property
+    def params(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self.delegate is not None
+            return self.delegate.params
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def distributions(self) -> Dict[str, BaseDistribution]:
+        def func() -> Dict[str, BaseDistribution]:
+
+            assert self.delegate is not None
+            return self.delegate.distributions
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def user_attrs(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self.delegate is not None
+            return self.delegate.user_attrs
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def system_attrs(self) -> Dict[str, Any]:
+        def func() -> Dict[str, Any]:
+
+            assert self.delegate is not None
+            return self.delegate.system_attrs
+
+        return _call_and_communicate_obj(func)
+
+    @property
+    def datetime_start(self) -> Optional[datetime]:
+        def func() -> Optional[datetime]:
+
+            assert self.delegate is not None
+            return self.delegate.datetime_start
+
+        return _call_and_communicate_obj(func)
+
+
+def _call_and_communicate(func: Callable, dtype: "torch.dtype") -> Any:
+    buffer = torch.empty(1, dtype=dtype)
+    rank = dist.get_rank()
+    if rank == 0:
+        result = func()
+        buffer[0] = result
+    if dist.get_backend() == "nccl":
+        buffer = buffer.cuda(torch.device(rank))
```

Comment:
[Just a note] It seems that we only test with the `gool` backend, so we have not confirmed this one in the unit-tests. However, since GitHub Actions does not provide GPU environments, it cannot be confirmed by unit tests.